parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -212,8 +212,20 @@ private static MetadataFilter filter(boolean skipRowGroups) {
   }
 
   private static <T> List<T> runAllInParallel(int parallelism, List<Callable<T>> toRun) throws ExecutionException {
-    LOG.info("Initiating action with parallelism: {}", parallelism);
-    ExecutorService threadPool = Executors.newFixedThreadPool(parallelism);
+    if (toRun.isEmpty()) {
+      return Collections.emptyList();
+    } else if (toRun.size() == 1) {
```

Comment:
I don't think this is a safe optimization. Running something in the current thread vs. running it in a separate thread could lead to different behavior. I believe that even if you caught all Throwable-s, it could still behave differently, so I think the single task case should not be handled differently.